resources/tentative/todomvc/src/components/todo-app/todo-app.component.js

```diff
@@ -0,0 +1,171 @@
+
+        });
+
+        this.update("update-item", event.detail.id);
+    }
+
+    toggleItems(event) {
+        this.list.toggleItems(event.detail.completed);
+    }
+
+    clearCompletedItems() {
+        this.list.removeCompletedItems();
+    }
+
+    update(type = "", id = "") {
+        const totalItems = this._data.length;
+        const activeItems = this._data.filter((entry) => !entry.completed).length;
+        const completedItems = totalItems - activeItems;
+
+        this.list.setAttribute("total-items", totalItems);
+        this.list.updateElements(type, id);
+
+        this.topbar.setAttribute("total-items", totalItems);
+        this.topbar.setAttribute("active-items", activeItems);
+        this.topbar.setAttribute("completed-items", completedItems);
+
+        this.bottombar.setAttribute("total-items", totalItems);
+        this.bottombar.setAttribute("active-items", activeItems);
+    }
+
+    addListeners() {
+        this.topbar.addEventListener("toggle-all", this.toggleItems);
+        this.topbar.addEventListener("add-item", this.addItem);
+
+        this.list.addEventListener("toggle-item", this.toggleItem);
+        this.list.addEventListener("remove-item", this.removeItem);
+        this.list.addEventListener("update-item", this.updateItem);
+
+        this.bottombar.addEventListener(
+            "clear-completed-items",
+            this.clearCompletedItems
+        );
+    }
+
+    removeListeners() {
+        this.topbar.removeEventListener("toggle-all", this.toggleItems);
+        this.topbar.removeEventListener("add-item", this.addItem);
+
+        this.list.removeEventListener("toggle-item", this.toggleItem);
+        this.list.removeEventListener("remove-item", this.removeItem);
+        this.list.removeEventListener("update-item", this.updateItem);
+
+        this.bottombar.removeEventListener(
+            "clear-completed-items",
+            this.clearCompletedItems
+        );
+    }
+
+    routeChange = (route) => {
+        const routeName = route.split("/")[1] || "all";
+        this.list.updateRoute(routeName);
+        this.bottombar.updateRoute(routeName);
+        this.topbar.updateRoute(routeName);
+    };
+
+    connectedCallback() {
+        this.connected = true;
+        const node = document.importNode(template.content, true);
+
+        this.topbar = new TodoTopbar();
+        this.list = new TodoList();
+        this.bottombar = new TodoBottombar();
+
+        node.querySelector("[name=\"topbar\"]").append(this.topbar);
+        node.querySelector("[name=\"list\"]").append(this.list);
+        node.querySelector("[name=\"bottombar\"]").append(this.bottombar);
+
+        this.list.addItems(this._data);
+        this.addListeners();
+        this.update("connected");
+
+        this.shadow.append(node);
+
+        this.router.initRouter(this.routeChange);
+        this._isReady = true;
+    }
+
+    disconnectedCallback() {
+        this.connected = false;
+        this.removeListeners();
```

Comment:
What's the point of removing event listener here?

Reply:
In case the app ever gets removed from the DOM, which we don't need here. 
I can remove

Reply:
disconnectedCallback in all components will now be responsible to remove event listeners that are added in the connectedCallback.